### PR TITLE
Adding continuous time value-iteration ADP method

### LINKED ISSUE #18: Add an ADP example
Implement nonlinear continuous-time (CT) value-iteration (VI) adaptive dynamic programming (ADP) algorithm [1].

### Refs.
[1] T. Bian and Z.-P. Jiang, “Value Iteration, Adaptive Dynamic Programming, and Optimal Control of Nonlinear Systems,” in 2016 IEEE 55th Conference on Decision and Control (CDC), Las Vegas, NV, USA, Dec. 2016, pp. 3375–3380, doi: 10.1109/CDC.2016.7798777.

- Note: [Reference code](https://github.com/tao-bian-nyu/nonlinear-VI).

### PR DESCRIPTION
**Pull request by JinraeKim**
# Summary
Continuous-time (CT) value-iteration (VI) adaptive dynamic programming (ADP; a.k.a. approximate DP) is implemented [1].
Some parameters are borrowed from the previous code written by T. Bian (in `R`) [2].

# Notes
## How to obtain the infimum (minimum)?
Note that the infimum of the approximate Hamiltonian [1, Eq. (16)] and the approximate optimal input [1, Thm. 2] are obtained by numerical optimisation in this implementation. On the other hand, the original implementation [2] relies on the structure of a specific function approximator to perform the optimisation.

This is a (potentially critical) limitation of the original paper.

## How to run the code?
Run `main/continuous_time_vi_adp.jl`.
### An example of simulation result
- Training progress
```julia
adp.V̂.param = [0.0, 0.0, 0.0]
i = 1
adp.V̂.param = [0.6991429619303489; 0.4005297738730431; 0.14493613208036282]
adp.V̂(x) = Polynomial{true, Float64}[0.6991429619303489x₁² + 0.4005297738730431x₁x₂ + 0.14493613208036282x₂²]
i = 2
adp.V̂.param = [0.8777618743585718; 0.40893790123571194; 0.23829955870488628]
adp.V̂(x) = Polynomial{true, Float64}[0.8777618743585718x₁² + 0.40893790123571194x₁x₂ + 0.23829955870488628x₂²]
i = 3
adp.V̂.param = [0.9396217654811783; 0.3500294314726259; 0.32908420114968484]
adp.V̂(x) = Polynomial{true, Float64}[0.9396217654811783x₁² + 0.3500294314726259x₁x₂ + 0.32908420114968484x₂²]
i = 4
adp.V̂.param = [0.9689124227258321; 0.283789244824134; 0.4226847336535787]
adp.V̂(x) = Polynomial{true, Float64}[0.9689124227258321x₁² + 0.283789244824134x₁x₂ + 0.4226847336535787x₂²]
i = 5
adp.V̂.param = [0.98724708467294; 0.22191049831107348; 0.5172381029219096]
adp.V̂(x) = Polynomial{true, Float64}[0.98724708467294x₁² + 0.22191049831107348x₁x₂ + 0.5172381029219096x₂²]
i = 6
adp.V̂.param = [1.0008302225178511; 0.166501881304451; 0.6093622899080263]
adp.V̂(x) = Polynomial{true, Float64}[1.0008302225178511x₁² + 0.166501881304451x₁x₂ + 0.6093622899080263x₂²]
i = 7
adp.V̂.param = [1.0112731880714034; 0.11828864573556205; 0.6952782075797261]
adp.V̂(x) = Polynomial{true, Float64}[1.0112731880714034x₁² + 0.11828864573556205x₁x₂ + 0.6952782075797261x₂²]
i = 8
adp.V̂.param = [1.018699323879966; 0.07816716984803262; 0.7714270710492228]
adp.V̂(x) = Polynomial{true, Float64}[1.018699323879966x₁² + 0.07816716984803262x₁x₂ + 0.7714270710492228x₂²]
i = 9
adp.V̂.param = [1.0229635945512883; 0.04694227707866673; 0.8352098668933136]
adp.V̂(x) = Polynomial{true, Float64}[1.0229635945512883x₁² + 0.04694227707866673x₁x₂ + 0.8352098668933136x₂²]
i = 10
adp.V̂.param = [1.0242378402667032; 0.024679553235220295; 0.8856126280041667]
adp.V̂(x) = Polynomial{true, Float64}[1.0242378402667032x₁² + 0.024679553235220295x₁x₂ + 0.8856126280041667x₂²]
i = 11
adp.V̂.param = [1.0231516780659369; 0.01040933897948123; 0.923321580504069]
adp.V̂(x) = Polynomial{true, Float64}[1.0231516780659369x₁² + 0.01040933897948123x₁x₂ + 0.923321580504069x₂²]
i = 12
adp.V̂.param = [1.0205905389583578; 0.0024068450854986614; 0.9502476179259722]
adp.V̂(x) = Polynomial{true, Float64}[1.0205905389583578x₁² + 0.0024068450854986614x₁x₂ + 0.9502476179259722x₂²]
i = 13
adp.V̂.param = [1.0173925637241983; -0.001240126660707691; 0.9687872753732119]
adp.V̂(x) = Polynomial{true, Float64}[1.0173925637241983x₁² - 0.001240126660707691x₁x₂ + 0.9687872753732119x₂²]
i = 14
adp.V̂.param = [1.0141606607135396; -0.002178849613238054; 0.9812189081177559]
adp.V̂(x) = Polynomial{true, Float64}[1.0141606607135396x₁² - 0.002178849613238054x₁x₂ + 0.9812189081177559x₂²]
i = 15
adp.V̂.param = [1.011234784530039; -0.0016124251794424668; 0.9894000097929393]
adp.V̂(x) = Polynomial{true, Float64}[1.011234784530039x₁² - 0.0016124251794424668x₁x₂ + 0.9894000097929393x₂²]
stop_conds_dict = Dict{Symbol, Bool}(:max_iter => 0, :w_tol => 1)

```
- After training
image

# Notice
The algorithm seems poorly work with other basis function, which is not trivial to obtain the optimiser of approximate Hamiltonian and optimal input.
It will be checked with more clear test codes.

# References
[1] [T. Bian and Z.-P. Jiang, “Value Iteration, Adaptive Dynamic Programming, and Optimal Control of Nonlinear Systems,” in 2016 IEEE 55th Conference on Decision and Control (CDC), Las Vegas, NV, USA, Dec. 2016, pp. 3375–3380. doi: 10.1109/CDC.2016.7798777.](https://ieeexplore.ieee.org/document/7798777)
[2] [https://github.com/JinraeKim/nonlinear-VI](A forked repo of the implemented code by T. Bian)

### CODE REVIEW

**Comment by JinraeKim** [comment]
@hnlee77
Welcome to review this PR. Please see the description of this PR.
Note that this PR would be merged before your review.

**Review by hnlee77 — APPROVED** [review]

File: src/environments/basics/TwoDimensionalNonlinearPolynomialEnv.jl [hunk 1]
```diff
@@ -10,9 +10,11 @@ end
 
 function dynamics!(env::TwoDimensionalNonlinearPolynomialEnv)
     return function (dx, x, p, t; u)
+        @assert length(u) == 1
+        _u = u[1]  # Real or Array
```
Comment (hnlee77) [inline src/environments/basics/TwoDimensionalNonlinearPolynomialEnv.jl:14]: Why did you do this?

File: src/algorithms/ADP.jl [hunk 2]
```diff
@@ -65,30 +82,48 @@ function set_data!(adp::CTValueIterationADP, data_new)
...
+"""
+function approximate_optimal_input(adp::CTValueIterationADP)
+    # TODO: Add inexact way of obtaining approximate optimal input;
+    # see https://github.com/JinraeKim/nonlinear-VI/blob/main/nonlinear_sys.R#L182
+    ŵ_f = adp.V̂.param
+    ĉ_f = adp.Θ * ŵ_f
+    return function (x, p, t)
+        û = min_Ĥ(adp)(x, ĉ_f)[2]
+    end
+end
```
Comment (hnlee77) [inline src/algorithms/ADP.jl:129]: Is this not used yet?

File: src/environments/basics/TwoDimensionalNonlinearPolynomialEnv.jl (hunk 1 above)
Comment (JinraeKim) [reply src/environments/basics/TwoDimensionalNonlinearPolynomialEnv.jl:14]: TL; DR) For convenient usage (while losing type stability).

In Julia, the types of `1.0` (`Number`) and `[1.0]` (`Array`) are different.
The dynamical equations of the system considered are based on control input whose type is `Number`.
On the other hand, the optimisation solver (`NLopt`) produces optimizer as an array whose length is `1`.
Hence, for convenience, the lines you highlighted make it possible to receive the control input of types both `Number` or `Array`.

File: src/algorithms/ADP.jl (hunk 2 above)
Comment (JinraeKim) [reply src/algorithms/ADP.jl:129]: See [here](https://github.com/JinraeKim/FlightSims.jl/pull/8/files#diff-8908bd017b8d9c0a61bbd2a586d6a1776985ccf2805d9ae6cb8084cff170b6a8R98).

**Review by hnlee77 — APPROVED** [review]

**Pull request merged**